rust/emitter/src/reference_op_emitter.rs

```diff
@@ -0,0 +1,501 @@
+use super::emitter::{AtomIndex, EmitError};
+use crate::ast_emitter::AstEmitter;
+
+#[derive(Debug, PartialEq)]
+enum SetOpReferenceKind {
+    GlobalName(AtomIndex),
+    #[allow(dead_code)]
+    Prop(AtomIndex),
+    #[allow(dead_code)]
+    Elem,
+}
+
+// See SetOpReferenceEmitter.
+// Use struct to hide the details from the consumer.
+#[derive(Debug)]
+#[must_use]
+pub struct SetOpReference {
+    kind: SetOpReferenceKind,
+}
+impl SetOpReference {
+    fn new(kind: SetOpReferenceKind) -> Self {
+        Self { kind }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum CallOpKind {
+    Normal,
+    // FIXME: Support eval, call, apply etc.
+}
+
+// See CallOpReferenceEmitter.
+// Use struct to hide the details from the consumer.
+#[derive(Debug)]
+#[must_use]
+pub struct CallOpReference {
+    kind: CallOpKind,
+}
+impl CallOpReference {
+    fn new(kind: CallOpKind) -> Self {
+        Self { kind }
+    }
+}
+
+// Struct for emitting bytecode for get operation.
+//
+// Usage:
+//
+//   `name;`
+//      GetEmitter::emit_name(self, name);
+//
+//   `obj.key;`
+//      GetEmitter::emit_prop(self, |emitter| {
+//          emitter.emit_expression(obj)
+//      }, key)?;
+//
+//   `obj[key];`
+//      GetEmitter::emit_elem(self, |emitter| {
+//          emitter.emit_expression(obj)
+//      }, |emitter| {
+//          emitter.emit_expression(key)
+//      })?;
+//
+pub struct GetEmitter {}
+impl GetEmitter {
+    pub fn emit_name(emitter: &mut AstEmitter, name: &str) {
+        let name_index = emitter.emit.get_atom_index(name);
+
+        //              [stack]
+
+        // FIXME: Support non-global case.
+        emitter.emit.get_g_name(name_index);
+        //              [stack] VAL
+    }
+
+    pub fn emit_prop<F>(
+        emitter: &mut AstEmitter,
+        obj_emitter: F,
+        key: &str,
+    ) -> Result<(), EmitError>
+    where
+        F: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+    {
+        let key_index = emitter.emit.get_atom_index(key);
+
+        //              [stack]
```

Comment:
I don't understand why you put so much space there. I would just remove all leading spaces.

Reply:
I've aligned them to column=24, so that they can be aligned even inside indent.
(I expect there can be 1 or 2 more indentation, depending on scope things.

Reply:
maybe better adding pre-commit hook to align them.
there's a script that does so for C++ (invoked manually, tho)
https://searchfox.org/mozilla-central/source/js/src/frontend/align_stack_comment.py